newlib/libc/include/errno.h

```diff
@@ -36,9 +36,11 @@ SUCH DAMAGE.
 
 #ifndef __error_t_defined
 typedef int error_t;
+
 #define __error_t_defined 1
 #endif
 
 #include <sys/errno.h>
+#include <sys/_types.h>
```

Comment:
This doesn't appear to be necessary in this file?

Reply:
Ah, I see the issue. The C standard says that errno.h defines `errno_t` if `__STDC_WANT_LIB_EXT1__` is defined before the inclusion of errno.h. Further, `stdlib.h` defines the various constraint handler functions to take parameters of type `errno_t`. But, it does not say that stdlib.h defines `errno_t`, so we must not define it in that case.

Here's what we do to make that work in picolibc by creating an internal type, `__errno_t`, so that the public `errno_t` type doesn't appear except in the correct situation:

Define `__errno_t` in sys/_types.h:
```
typedef int __errno_t;
```

Include sys/_types (to get __errno_t defined) and define `errno_t` in errno.h when it should be:
```
#include <sys/_types.h>
#ifdef __STDC_WANT_LIB_EXT1__
typedef __errno_t errno_t;
#endif
```

Use `__errno_t` in sys/errno.h instead of the bare 'int' used currently. Note *do not use `errno_t`* as that may not be defined at all. 

Use `__errno_t` in stdlib.h when defining the constraint_handler functions.
```
#ifdef __STDC_WANT_LIB_EXT1__
#include <sys/_types.h>
...
void abort_handler_s(const char * restrict msg, void * restrict ptr,
__errno_t error);
...
#endif
```

The requirement here is that `errno_t` must be define if you include errno.h with `__STDC_WANT_LIB_EXT1__` defined, otherwise it must not be defined. You must also be able to include stdlib.h without having included errno.h, even if `__STDC_WANT_LIB_EXT1__` is defined, and in that case, `errno_t` must not be defined, *but* the constraint_handler API must be declared, even though you cannot use it unless you also include errno.h (as the `errno_t` type won't be defined).